### DISCUSSION TITLE
FileUpload - File Extension Change via livewire/upload-file

### DISCUSSION THREAD

**Discussion by falconsmilie** [filamentphp/filament#6528] [Help] [answered]
Hi. I'm not sure where this belongs (Filament/Livewire), but our pen testers have found an interesting thing with the FileUpload field where a user can use `livewire/upload-file` to change the extension of the file being uploaded.

Our intention is to allow users to upload text files (.txt). The FileUpload field looks like this;

```
FileUpload::make('downloads')
            ->multiple()
            ->directory('my_download_directory')
            ->disableLabel()
            ->acceptedFileTypes([
                'txt/plain',
            ])
            ->enableReordering()
            ->preserveFilenames();
```

The validation on this field can be ignored by changing the extension of a text file (`test1234.txt`) while uploading the file via `/livewire/upload-file`.

```
POST /livewire/upload-file?expires=1684246695&signature=0013e3fc11fbfc5263d5de366c443106139bcea1fb6e49221fe50b00ba562488 HTTP/2
Host: myhost.com

...SNIP...

------WebKitFormBoundaryVdD5p9fLcFYtD6Rp
Content-Disposition: form-data; name="files[]"; filename="test1234.php"
Content-Type: text/plain

test
<?php
echo "test123";
?>

------WebKitFormBoundaryVdD5p9fLcFYtD6Rp--
```

Any suggestions for how to prevent this problem?

**Comment by danharrin** [comment]
If you check out the `acceptedFileTypes()` internally, it actually adds a Laravel `mimetypes` rule. I think that since the mime type is still plain text (?) then it passes the validation? Changing the extension doesnt change the mime, right?

**Reply by falconsmilie** [reply]
Yeah, the mime type remains the same. 

I looked at `acceptedFileTypes()` and I tried the laravel `mimetypes `validation earlier, but it is the same result. As you say, `acceptedFileTypes` and Laravel's `mimetypes` validation are basically the same logic.

The issue i have is with the extension being changed. txt to php is, as you can imagine, not a good thing in this case.

**Reply by danharrin** [reply]
Right, but if you are using something like S3 it really shouldn't be a problem. Just make sure you aren't using your server's filesystem?

**Reply by falconsmilie** [reply]
So many questions :D There is no way to prevent the extension renaming from happening? It seems `getUploadedFileNameForStorageUsing()` and `saveUploadedFileUsing()` cannot help in this instance. How about the Livewire config temporary upload rules, or will that not help either?

**Reply by danharrin** [reply, marked as answer]
Unless you can apply a validation rule for the file extension? But I don't think so. As I said, I don't think its a problem if you use S3

**Reply by falconsmilie** [reply]
My attacker would need to rename to `.html` instead, and add some malicious JS in there, instead of PHP? 

Not sure how S3 serves html if it's not setup for a static website.

**Reply by falconsmilie** [reply]
Adding a file extension validator on the field's rules is preventing the renaming from being successfully executed. Thanks for your help Dan.
